### PR TITLE
KOGITO-1631: file system based persistence addon

### PR DESCRIPTION
**Pull request by mswiderski**

### CODE REVIEW

**Comment by mswiderski** [comment]
similar as last time ... can someone share what is the reason of the build failure?

**Comment by cristianonicolai** [comment]
@mswiderski sonar is failing due to:
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-antrun-plugin:1.8:run (default) on project kogito-runtimes: An Ant BuildException has occured: Error while creating report

[ERROR] around Ant part ...<report>... @ 7:11 in /home/jenkins/workspace/IE_kogito_kogito-runtimes_PR-400/target/antrun/build-main.xml: Error while analyzing addons/persistence/infinispan-persistence-addon/target/classes/org/kie/kogito/persistence/KogitoProcessInstancesFactory.class. Can't add different class with same name: org/kie/kogito/persistence/KogitoProcessInstancesFactory

File: addons/persistence/filesystem-persistence-addon/.gitignore [hunk 1]
```diff
@@ -0,0 +1,10 @@
+target/
```
Comment (cristianonicolai) [inline addons/persistence/filesystem-persistence-addon/.gitignore]: shouldnt need to add, same rules should be handled by root .gitignore

File: addons/persistence/filesystem-persistence-addon/src/main/java/org/kie/kogito/persistence/KogitoProcessInstancesFactory.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
```
Comment (cristianonicolai) [inline addons/persistence/filesystem-persistence-addon/src/main/java/org/kie/kogito/persistence/KogitoProcessInstancesFactory.java]:
```suggestion
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
```
:D

File: addons/persistence/filesystem-persistence-addon/src/main/java/org/kie/kogito/persistence/filesystem/FileSystemProcessInstances.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+	protected void disconnect(Path processInstanceStorage, ProcessInstance instance) {
+		((AbstractProcessInstance<?>) instance).internalRemoveProcessInstance(() -> {
+
+			try {
+				byte[] reloaded = readBytesFromFile(processInstanceStorage);
+
+	            return ((AbstractProcessInstance<?>)marshaller.unmarshallProcessInstance(reloaded, process, (AbstractProcessInstance<?>) instance)).internalGetProcessInstance();
+			} catch (RuntimeException e) {
+				return null;
```
Comment (cristianonicolai) [inline addons/persistence/filesystem-persistence-addon/src/main/java/org/kie/kogito/persistence/filesystem/FileSystemProcessInstances.java]: should this be propagated?

File: addons/persistence/filesystem-persistence-addon/src/main/java/org/kie/kogito/persistence/filesystem/FileSystemProcessInstances.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        
+        UserDefinedFileAttributeView view = Files.getFileAttributeView(file, UserDefinedFileAttributeView.class);
+        if (view != null) {
+            try {
+                java.nio.ByteBuffer bb = java.nio.ByteBuffer.allocate(view.size(key));
+                view.read(key, bb);
+                bb.flip();
+                return Charset.defaultCharset().decode(bb).toString();
+            } catch (IOException e) {
+                return null;
```
Comment (cristianonicolai) [inline addons/persistence/filesystem-persistence-addon/src/main/java/org/kie/kogito/persistence/filesystem/FileSystemProcessInstances.java]: same here, should error be propagated?

File: addons/persistence/filesystem-persistence-addon/src/main/java/org/kie/kogito/persistence/filesystem/FileSystemProcessInstances.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+	            return ((AbstractProcessInstance<?>)marshaller.unmarshallProcessInstance(reloaded, process, (AbstractProcessInstance<?>) instance)).internalGetProcessInstance();
+			} catch (RuntimeException e) {
+				return null;
+			}
+
+
+        });
+	}
+	
+	protected String getMetadata(Path file, String key) {
```
Comment (cristianonicolai) [inline addons/persistence/filesystem-persistence-addon/src/main/java/org/kie/kogito/persistence/filesystem/FileSystemProcessInstances.java]: seems that metatada is set but never read, should it be checked, maybe a test?

**Comment by evacchi** [comment]
shall we open JIRAs for contributions as well?

**Comment by cristianonicolai** [comment]
> shall we open JIRAs for contributions as well?

+1

File: addons/persistence/filesystem-persistence-addon/.gitignore (hunk 1 above)
Comment (mswiderski) [reply addons/persistence/filesystem-persistence-addon/.gitignore]: you mean to remove the .gitignore completely, right?

File: addons/persistence/filesystem-persistence-addon/src/main/java/org/kie/kogito/persistence/filesystem/FileSystemProcessInstances.java (hunk 4 above)
Comment (mswiderski) [reply addons/persistence/filesystem-persistence-addon/src/main/java/org/kie/kogito/persistence/filesystem/FileSystemProcessInstances.java]: this one no, as it optional to store the meta data and depends actually on file system being used.

File: addons/persistence/filesystem-persistence-addon/src/main/java/org/kie/kogito/persistence/filesystem/FileSystemProcessInstances.java (hunk 5 above)
Comment (mswiderski) [reply addons/persistence/filesystem-persistence-addon/src/main/java/org/kie/kogito/persistence/filesystem/FileSystemProcessInstances.java]: can't be added as reliable test as this depends on file system that might not support the user defined file attributes view so consider it a optional and not required for the actual persistence to work. More of a utility to be able to identify the process instance on certain file systems.

**Comment by mswiderski** [comment]
here is jira https://issues.redhat.com/browse/KOGITO-1631

File: addons/persistence/filesystem-persistence-addon/src/main/java/org/kie/kogito/persistence/filesystem/FileSystemProcessInstances.java (hunk 3 above)
Comment (mswiderski) [reply addons/persistence/filesystem-persistence-addon/src/main/java/org/kie/kogito/persistence/filesystem/FileSystemProcessInstances.java]: this is called from *ProcessInstance whenever it needs to load the process from the storage. It relies on returned null and then produces ProcessInstanceNotFoundException. So I will all logging of the exception to indicate root cause of the process instance not found.

**Comment by mswiderski** [comment]
closing for the time being as there is a feature freeze

**Pull request closed without merging**